From 218ef5ef52c2a8b2326d063460b1cfb414e102cc Mon Sep 17 00:00:00 2001
From: James Chapman <jachapma@redhat.com>
Date: Mon, 6 Oct 2025 12:32:08 +0100
Subject: [PATCH] Issue 7027 - 389-ds-base OpenScanHub Leaks Detected (#7028)

Description: A file is opened using open() and then wrapped in a
zlib gzFile with gzdopen(). If gzdopen() fails, it never takes
ownership of the fd, so it remains open.

Fix: Ensure fd is explicitly closed when gzdopen() returns NULL.

Fixes: https://github.com/389ds/389-ds-base/issues/7027

Reviewed by: @mreynolds389  (Thank you)
---
 ldap/servers/slapd/log.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/ldap/servers/slapd/log.c b/ldap/servers/slapd/log.c
index 58f9fb4d6..668a02454 100644
--- a/ldap/servers/slapd/log.c
+++ b/ldap/servers/slapd/log.c
@@ -191,8 +191,12 @@ compress_log_file(char *log_name, int32_t mode)
      * compressed content.
      */
     if ((fd = open(gzip_log, O_WRONLY|O_CREAT|O_TRUNC, mode)) >= 0) {
-        /* FIle successfully created, now pass the FD to gzdopen() */
+        /* File successfully created, now pass the FD to gzdopen() */
         outfile = gzdopen(fd, "ab");
+        if (outfile == NULL) {
+            close(fd);
+            return -1;
+        }
     } else {
         return -1;
     }
-- 
2.52.0

